File: Java+/UnitTest+/OkHttpMockWebServer/test/server/HelloWorld.java
package server;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import org.junit.jupiter.api.Test;
import util.NetUtil;

import java.io.IOException;
import java.net.URL;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class HelloWorld {

    @Test
    void mock() throws IOException, InterruptedException {
        var server = new MockWebServer();

        var body1 = "hello, world!";
        var body2 = "sup, bra?";
        var body3 = "yo dog";
        server.enqueue(new MockResponse().setBody(body1));
        server.enqueue(new MockResponse().setBody(body2));
        server.enqueue(new MockResponse().setBody(body3));

        server.start();

        var baseUrl = server.url("/v1/chat/");

        var body = NetUtil.urlContentToString(new URL(baseUrl.url(), "messages/"));
        assertThat(body, equalTo(body1));

        var actBody2 = NetUtil.urlContentToString(new URL(baseUrl.url(), "messages/2"));
        assertThat(actBody2, equalTo(body2));

        var actBody3 = NetUtil.urlContentToString(new URL(baseUrl.url(), "messages/3"));
        assertThat(actBody3, equalTo(body3));

        var request1 = server.takeRequest();
        assertEquals("/v1/chat/messages/", request1.getPath());
        assertThat(request1.getMethod(), equalTo("GET"));

        var request2 = server.takeRequest();
        assertEquals("/v1/chat/messages/2", request2.getPath());

        var request3 = server.takeRequest();
        assertEquals("/v1/chat/messages/3", request3.getPath());

        server.shutdown();
    }
}

File: Scala+/ScalaSbt212/Libs+/OkHttpMockWebServer/src/test/scala/server/HelloWorld.scala
package server

import okhttp3.mockwebserver.{MockResponse, MockWebServer}
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper

import java.net.URL
import scala.io.Source


class HelloWorld extends AnyFlatSpec {

  "Mock Server" should "return responses" in {
    val server = new MockWebServer

    val body1 = "hello, world!"
    val body2 = "sup, bra?"
    val body3 = "yo dog"
    server.enqueue(new MockResponse().setBody(body1))
    server.enqueue(new MockResponse().setBody(body2))
    server.enqueue(new MockResponse().setBody(body3))

    server.start()

    val baseUrl = server.url("/v1/chat/")

    val source1 = Source.fromURL(new URL(baseUrl.url, "messages/"))
    val actBody1 = source1.mkString
    source1.close()
    actBody1 shouldBe body1

    val source2 = Source.fromURL(new URL(baseUrl.url, "messages/2"))
    val actBody2 = source2.mkString
    source2.close()
    actBody2 shouldBe body2

    val source3 = Source.fromURL(new URL(baseUrl.url, "messages/3"))
    val actBody3 = source3.mkString
    source3.close()
    actBody3 shouldBe body3

    val request1 = server.takeRequest
    request1.getPath shouldBe "/v1/chat/messages/"
    request1.getMethod shouldBe "GET"

    val request2 = server.takeRequest
    request2.getPath shouldBe "/v1/chat/messages/2"

    val request3 = server.takeRequest
    request3.getPath shouldBe "/v1/chat/messages/3"

    server.shutdown()
  }
}
